File: kernel/cmd/tempora-studio-host/packagegrants.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"

	"tempora/internal/platform/packagegrant"
)

// stripPackageGrants works on the directory the application runs from and is
// never handed a directory directly: the grants it removes are the ones that
// stop that application's own sandboxed children from loading, and no other
// tree is this flag's to change. The report goes to out as one JSON line.
func stripPackageGrants(out, logs io.Writer, app string) int {
	if strings.TrimSpace(app) == "" {
		fmt.Fprintln(logs, "strip-package-grants: -studio-app names no application")
		return 2
	}
	exe, err := filepath.EvalSymlinks(app)
	if err != nil {
		fmt.Fprintf(logs, "strip-package-grants: %v\n", err)
		return 2
	}
	if info, err := os.Stat(exe); err != nil || !info.Mode().IsRegular() {
		fmt.Fprintf(logs, "strip-package-grants: %s is not an application executable\n", exe)
		return 2
	}
	report, err := packagegrant.Strip(filepath.Dir(exe))
	if err != nil {
		fmt.Fprintf(logs, "strip-package-grants: %v\n", err)
		return 1
	}
	for _, refused := range report.Refused {
		fmt.Fprintf(logs, "strip-package-grants: kept a grant at %s: %v\n", refused.Path, refused.Err)
	}
	for _, unread := range report.Unread {
		fmt.Fprintf(logs, "strip-package-grants: could not read %s: %v\n", unread.Path, unread.Err)
	}
	if err := json.NewEncoder(out).Encode(report); err != nil {
		return 1
	}
	return 0
}

File: kernel/desktop/electron/src/packagegrants.js
"use strict";
const { execFileSync } = require("node:child_process");

const SHOWN_PATHS = 5;

// Chromium's sandboxed children exit while loading a DLL whose access entries
// grant one Windows app package, so the window never paints. The kernel removes
// those grants from the tree this executable runs from, and it has to happen
// before Chromium starts its first child — which is why this runs synchronously.
function stripPackageGrants(binary, { platform, packaged, execPath }, run = execFileSync) {
  if (platform !== "win32" || !packaged) return null;
  try {
    const raw = run(binary, ["-strip-package-grants", "-studio-app", execPath], {
      encoding: "utf8",
      timeout: 15000,
      windowsHide: true,
      // The kernel names each object it kept a grant on or could not read there.
      stdio: ["ignore", "pipe", "inherit"],
    });
    return readReport(raw);
  } catch (err) {
    console.error("tempora-studio: strip-package-grants:", err.message);
    return null;
  }
}

function readReport(raw) {
  const body = JSON.parse(raw);
  if (!Array.isArray(body?.stripped) || !Array.isArray(body?.refused)) {
    throw new Error("the grant report carried no lists");
  }
  const stripped = body.stripped.filter((p) => typeof p === "string");
  const refused = body.refused.map((r) => r?.path).filter((p) => typeof p === "string");
  return { stripped, refused };
}

// What to tell someone whose window died before it painted, when the kernel
// already knows why: grants it found and could not remove. Anything else is a
// cause this shell does not know, and it says nothing rather than guess.
function unpaintedWindowCause(report, locale) {
  if (!report?.refused.length) return null;
  const shown = report.refused.slice(0, SHOWN_PATHS).join("\n");
  const more = report.refused.length - SHOWN_PATHS;
  if (String(locale).toLowerCase().startsWith("zh")) {
    return {
      title: "Tempora Studio 无法打开窗口",
      detail:
        "Tempora Studio 加载的文件上有授予某个 Windows 应用包（AppContainer）的访问项，Chromium 的沙箱进程加载这类文件时会退出。" +
        "Studio 尝试移除它们，以下位置被拒绝：\n\n" + shown + (more > 0 ? `\n……另有 ${more} 处` : "") +
        "\n\n以管理员身份运行一次 Tempora Studio，它会自行移除这些访问项。",
    };
  }
  return {
    title: "Tempora Studio could not open its window",
    detail:
      "Files Tempora Studio loads carry access entries granting a specific Windows app package (AppContainer), " +
      "and Chromium's sandboxed processes exit while loading such a file. Studio tried to remove them and was refused for:\n\n" +
      shown + (more > 0 ? `\n…and ${more} more` : "") +
      "\n\nRun Tempora Studio once as administrator and it will remove them itself.",
  };
}

module.exports = { stripPackageGrants, readReport, unpaintedWindowCause };
